### ISSUE TITLE
[Feature Request] Escaping interpreted characters in numbering patterns

### ISSUE THREAD

**Issue by edhebi** [typst/typst#1177]
### Description

A number of characters are interpreted in numbering patterns (since that's the whole point). Because there's no escape hatch, this makes them impossible to use in the various prefix of suffix. I propose that a backlash character should stop the next character from being interpreted in this way. For example a patter of `"Section 1.1 )"` would be written `"Sect\\ion 1.1 )"`

I can see two limitations:

* It's not obvious that you need a double backslash instead of of a single one to someone not familiar with escape sequences, and that makes an actual backslash be `"\\\\"` but I suspect that's not a common character at all to put in there.
* It's added complexity for something that can currently be solved with numbering functions. I do think this makes a bunch of patterns nicer for the user on the long run tho.

I'd actually be willing to implement this if there's interested, that'd be a good way to get familiar with the codebase :p

### Use Case

For more involved numbering patterns, it can get frustrating to work around interpreted characters, especially `aAiI`. For example, `numbering("Section 1.1 )")` doesn't produce the expected result because the `i` is a special character in a numbering pattern.

With this feature it would be `numbering("Sect\\ion  1.1 )")` instead.

```c
// Before
#set heading(numbering: (..nums) => "Section " + numbering("1.1 )", ..nums.pos()))
// After
#set heading(numbering: "Sect\\ion 1.1 )")
```

**Comment by laurmaedje** [comment]
This is certainly an issue although I think having "Section" be part of the numbering is misuse. The "Section" part is a supplement and should be handled separately. 

I am not really convinced by the backslash solution though. It doesn't fix the problem that we cannot add new numbering kinds without it being a breaking change and, to be honest, it's just not particularly nice. 

A clean solution would of course be to "escape" the actual numberings (i.e. `"{1} / {1}"` or maybe with just a single letter), but that would make normal simple patterns so complicated. I am unsure how to proceed here.

We also have similar patterns for dates soon and there has been discussion about patterns for citations. We should find a general solution for this that is consistent across all these usages.

**Comment by edhebi** [comment]
supplement aren't really related to how the header/thing is displayed tho, if I'm understanding correctly. Right now, displaying something *before* the number in a heading requires a number prefix. My example of the word "Section" was a bit weird I'll admit; but being to reuse headings for other things in documents is useful. I had the case of doing homework sheets and trying to number the things "Exercice n -- [the heading here]".

I realize this isn't particularly elegant, but I don't think this is an issue that's solved in any way by supplements, unless I'm wildly missing what they can do

**Comment by laurmaedje** [comment]
True, supplements don't really solve that. But your solution has the problem that the supplement _plus_ the full numbering (including "Section") will be shown when referencing the heading, so section would show twice. To add "Exercise" before every heading, you'd need to write a custom heading show rule for now.

**Comment by bluebear94** [comment]
What about interpreting common numbering items such as `1` and `i` only when no `{}` escape is found?

**Comment by kamasylvia** [comment]
Is redefine `numbering` possible?
```
numbering(
  string/function,
  ..integer,
) -> any
```
To
```
numbering(
  string/function,
  prefixes: string/none,
  suffixes: string/none,
  ..integer,
) -> any
```

**Comment by edhebi** [comment]
I guess the issue could either be closed at this point or stay here as a question about string numbering limitations. I should also say that I wrote the numberingx package that solves a number of those issues, in particular requiring braces around special symbols. This also allows for pattern that aren't a single symbol, which makes internationalization much nicer.

**Comment by ElenaRgC** [comment]
I'm facing the same issue trying to make a heading with `#set heading(numbering: "Article 1:")`. Looking into the documentation, it doesn't seem to be a way to show that heading properly, and I think there needs to be a way to escape characters as common as vowels.

**Comment by jrmoulton** [comment]
Also facing the same issue.

**Comment by laurmaedje** [comment]
@jrmoulton You could also write a show rule for outline.entry. It's a bit of boilerplate, but conceptually cleaner.

**Comment by laurmaedje** [comment]
Related: https://github.com/typst/typst/issues/905

**Comment by laurmaedje** [comment]
> What about interpreting common numbering items such as 1 and i only when no {} escape is found?

I agree that this is the best way forward. I think the `{}` version should be there both for escaping and for more verbose names, so to get arabic numerals, there'd be three ways: `"1"`, `"{1}"`, and `"{arabic}"` and the latter two would prevent any free-standing numbering kinds from being interpreted.

To implement this, we would have to:
- Adjust `FromStr for NumberingPattern` to parse curly-braced identifiers, and if it finds one, disable free-standing numbering kinds (probably best done by keeping the current code for a first pass, and if we see a curly brace, restart parsing with a different function)
- Assign names to all numbering patterns. The `NumberingKind` enum already has names. We should check whether the (currently internal) names (but exposed as kebab-case) are good, or whether they can be made more uniform or e.g. aligned with CSS counter-styles.
- Adjust the IntoValue impl of NumberingPattern (`self => {` in `cast! { NumberingKind`) to always generate a valid thing. We can probably always output the braced version here for simplicity.

I should also note that @LuxxxLucy [expressed interest](https://github.com/typst/typst/pull/4273#issuecomment-2451422275) in implementing this.

**Comment by OverflowCat** [comment]
There is always a problem with using a single string. How can we separate the suffix and prefix? Currently, both `numbering: "1. (a)"`, `numbering: "1. - (a)"` yield `(1) (a)`. So why don't we define an array instead? We can even use a function in, say, the second level, and use str for those left. This will make writing custom numbering much more easier.

**Comment by Andrew15-5** [comment]
@OverflowCat, see https://github.com/typst/typst/issues/905.
